text/0000-bee-model-crate/0000-bee-model-crate.md

```diff
@@ -0,0 +1,334 @@
+This feature is responsible for the creation and interpretation of transactions and bundles.
+
+# Motivation
+
+IOTA is a distributed ledger that was designed for payment settlement and data transfer between machines and devices in the Internet of Things (IoT) ecosystem.
+The data packets that are sent through the network are called "transactions".
+Settlements or data transfers can be done with the help of these transactions. Payment settlements require the help of multiple transactions.
+
+# Detailed design
+
+## General
+
+A transaction consists of several fields (e.g. address, value, timestamp, tag). Each field is of static length. One transaction consists of 2673 trytes:
+
+- **trunk hash** the hash of the first transaction referenced/approved = 81 trytes
+- **branch hash** the hash of the second transaction referenced/approved = 81 trytes
+- **signature_and_message_fragment** contains the signature of the transfer or user-defined message data = 2187 trytes
+- **value** the transferred amount in IOTA  = 27 trytes
+- **address** receiver (output) if value > 0, or sender (input) if value < 0 = 81 trytes
+- **timestamp** the time when the transaction was issued = 9 trytes
+- **attachment_timestamp** the timestamp for when Proof-of-Work is completed = 9 trytes
+- **attachment_timestamp_lowerbound** is a slot for future use = 9 trytes
+- **attachment_timestamp_upperbound** is a slot for future use = 9 trytes
+- **tag** arbitrary user-defined value = 27 trytes
+- **obsolete_tag** another arbitrary user-defined tag = 27 trytes
+- **nonce** is the Proof-of-Work nonce of the transaction = 27 trytes
+- **bundle hash** is the hash of the entire bundle = 81 trytes
+- **current_index** the position of the transaction in its bundle = 9 trytes
+- **last_index** the total number of transactions in the bundle = 9 trytes
+
+A bundle is a collection of specific transactions. Bundles are required to bundle related information.
+For example, if not all data fits into one transaction, it has to be split across multiple transactions. An example would be payment settlements - they require the help of multiple transactions.
+A bundle then represents the collection of these transactions. It should be noted, that each transaction of a bundle will be sent separately through the network.
+
+Each transaction is identified by its hash, the transaction hash. Transactions should be final and therefore not modifiable after construction.
+The same applies to bundles. Bundles are final, and therefore not modifiable after construction.
+
+Transactions can be built from Transaction Builders.
+Bundles can be built from Bundle Builders. Both builder objects can be manipulated as desired.
+It should be noted, Bundle Builders don't process Transactions (as transactions are final), instead, they process Transaction Builders.
+
+## Exposed Interface
+
+### Transaction
+
+ A transaction consists of following fields:
+
+```rust
+struct Transaction {
+
+    signature_fragments: String,
```

Comment:
I'd like the name to reflect that it doesn't necessarily contain a signature. `signature_or_message` ?

Reply:
`Option` still allocates memory for the field so it's not really helpful

Reply:
@SuperFluffy what are your thoughts regarding `Option`. Semantically, it would be more correct, right?

Reply:
I'm not sure it makes more sense "semantically", an empty sig_or_msg doesn't mean there is no sig_or_msg. There is always a sig_or_msg :)

Reply:
Agree, so maybe Option isn't necessary. And if we use static types, we could initialize it with an empty (9's) `SignatureOrMessageFragment` and could be sure the internal state of the builder object is always valid.